SR/Code/sr/man/srm.1
.TH SRM 1 "15 December 1993" "University of Arizona"
.SH NAME
srm \- SR makefile generator
.SH SYNOPSIS
.B srm
[ options ] file.sr ...  [ other files and libraries ]
.SH DESCRIPTION
.LP
\fISrm\fP generates a makefile (for use with \fImake\fP(1)) that builds an
SR program from the given \fIsr\fP source files (\fI.sr\fP files),
other files (\fI.c\fP, \fI.o\fP files), libraries (-l\fIlib\fP), and 
archives (\fI.a\fP files).
.LP
\fISrm\fP produces a makefile that uses \fIsr\fP(1) and \fIsrl\fP(1) to
compile, link, and execute the program described by the files and libraries
listed on the command line.
The generated makefile provides the following services:
.TP 15
.B make compile
compiles without linking
.TP
.B make link
links after a ``make\ compile'' (really the same thing as just ``make'')
.TP
.B make run
executes the program (runtime arguments may be specified with the
\fB\-R\fP option)
.TP
.B make ls
produces a list of the source files (e.g., ``pr\ `make ls`'')
.TP
.B make clean
removes all artifacts of compilation except for the executable
.TP
.B make cleanx
removes all artifacts of compilation including the executable
(additional files may be specified for removal with the \fB\-Z\fP option)
.TP
.B make make
re-makes the makefile as per the original instructions (useful if the
resource import graph changes and the makefile needs to reflect the new
dependencies)
.LP
The default makefile name is
.BR Makefile ,
and thus the command
``make'' executes it.
The makefile name can be changed using the \fB\-f\fP option.
As a safety
feature, \fIsrm\fP refuses to overwrite an existing file that it did
not create.
.LP
The default executable name is
.BR a.out.
This can be changed using the 
.B \-o
option.
.LP
\fISrm\fP attempts to determine the main resource (for linking purposes)
by analyzing the source files and selecting the first resource it
finds that is imported by the least number of resources.
If some other resource is to be the main resource, it should be specified 
with the \fB\-m\fP option.
.LP
\fISrm\fP takes a number of option flags, which are
specified in arguments beginning with a hyphen.
The following options are recognized:
.TP 10
.BI \-f " mfile"
Name the makefile 
.I mfile
instead of
.BR Makefile .
A file name of ``\-'' directs the makefile to standard output.
.TP
.BI \-I " dir"
Use
.IR dir /Interfaces
as the Interfaces directory instead of ./Interfaces.
.TP
.BI \-o " ofile"
Use
.I ofile
for executable output instead of
.BR a.out .
.TP
.BI \-c " compiler"
Use
.I compiler 
instead of the default compiler.
.TP
.BI \-m " res"
Use the specified resource as the main resource.
The default is the first resource found that is imported
by the fewest number of resources.
.TP
.B \-v
Display interesting information while building the makefile.
.TP
.B \-w
Output wide makefile lines instead of limiting to 78 columns.
.TP
.BI \-C " opts"
Include
.I opts
in the list of options passed to the
compiler \fIsr\fP(1).
.TP
.BI \-L " opts"
Include
.I opts
in the list of options passed to the
linker \fIsrl\fP(1).
.TP
.BI \-R " args"
Include
.I args
in the list of arguments passed to the
executable by ``make\ run''.
.TP
.BI \-Z " files"
Include
.I files
in the list of files removed by
``make\ cleanx''.
.SH EXAMPLES
.TP 30
srm *.sr
no options, all .sr files in  current directory
.TP
srm \-m main \-o x a.sr b.sr
the main resource and executable name are specified
.TP
srm \-C "\-q \-O" a.sr b.sr 
flags are passed to \fIsr\fP(1) 
.TP
srm a.sr b.sr c.o \-ll r.c \-lm
objects, libraries and \fIC\fP sources listed
are passed on to \fIsrl\fP(1) to be linked in with the sr program.
.SH ENVIRONMENT
.IP SRMOPTS
The environment variable SRMOPTS may contain
.I srm
options to be processed ahead of explicit options.
.IP SR_PATH
.I Srm
searches for resources and object files
by looking first in the current directory,
then in the Interfaces directory,
and finally in the SR library.
(Items found somewhere other than the current directory are included in
the makefile as
.I srl
arguments but not as dependencies.)
If SR_PATH is set, any directories named there are checked
ahead of the standard search sequence.
Directories in SR_PATH are separated by colons (\fB:\fP).
.SH FILES
.ta 30n
Makefile	generated makefile
.br
file.sr		SR source file
.br
a.out	executable program
.br
Interfaces/component.o	object file
.br
Interfaces/resource.spec	export information
.SH SEE ALSO
sr(1), srl(1), make(1)
.SH DIAGNOSTICS
\fISrm\fP uses the \fIsr\fP compiler to obtain dependency information.
In addition to the syntax errors detected by the compiler,
.I srm
detects missing resource/global
specifications and duplicate spec/body definitions.  Warnings are 
issued if a resource or global has no body or a resource that is not the main
resource is not imported by any other resource. 
.SH CAVEATS
.PP
Omitting \fIsrm's\fP initial
.B s
can have disastrous results.
.PP
\fISrm\fP doesn't escape shell meta-characters in the makefile.
.PP
If the main resource is imported by others, the
.B \-m
option will probably be required to produce a correct makefile.


SR/Code/sr/srm/srm.c
#include <stdio.h>
#include "../config.h"
#include "../gen.h"
#include "../paths.h"
#include "../util.h"
#include "srm.h"

struct	symtabSt symtab[MAX_RES_DEF + 1];
int	numResources;

char	*compiler = NULL;		/* sr compiler to use */
char	*executable = "a.out";		/* name of the executable */
char	*makefileName = "Makefile";	/* name of the makefile */
char	*interfaceDir = NULL;		/* name of the interfaces dir */
char	*mainResName = NULL;		/* name of the main resource */
char	*compileOptions = "";		/* options to the sr compiler */
char	*linkerOptions = "";		/* options for srl */
char	*runtimeArgs = "";		/* arguments for the executable */
char	verbose = FALSE;		/* turns on/off srm verbose mode */
char	widelines = FALSE;		/* output wide lines? */
char	*zapFiles = NULL;		/* files deleted by make cleanx */
char	**sourceFiles;			/* sr source files processes by srm */
char	**otherSrc = NULL;		/* other source/binary files named */
char	**otherBin = NULL;		/* same, in .o form, NULL if no depend*/

/* main program */

main (argc, argv)
int	argc;
char	**argv;
{
    int		opt;
    extern	char *optarg;
    extern	int  optind;
    char	tempString[MAX_PATH];
    char	*getarg ();
    char	*env, *option;

    /* set exit message prefix */
    xprefix = argv[0];

    /* process options in environment SRMOPTS */
    if (env = getenv ("SRMOPTS"))
	while (option = strchr (env, '-')) {
	    switch (* (++option)) {
		case 'f':
		    makefileName = getarg (option);
		    break;
		case 'I':
		    interfaceDir = getarg (option);
		    break;
		case 'o':
		    executable = getarg (option);
		    break;
		case 'c':
		    compiler = getarg (option);
		    break;
		case 'm':
		    mainResName = getarg (option);
		    break;
		case 'v':
		    verbose = TRUE;
		    break;
		case 'w':
		    widelines = TRUE;
		    break;
		case 'C':
		    compileOptions = getarg (option);
		    break;
		case 'L':
		    linkerOptions = getarg (option);
		    break;
		case 'R':
		    runtimeArgs = getarg (option);
		    break;
		case 'Z':
		    zapFiles = getarg (option);
		    break;
		default:
		    usageMessage ();	/* does not return */
	    }
	    env = option;
	}

    /* process command line options */
    /* the "+" getopt arg works around gratuitous GNU getopt incompatibility */
    while ((opt = getopt (argc, argv, "+f:I:o:c:m:vwC:L:R:Z:")) != EOF)
	switch (opt) {
	    case 'f':
		makefileName = optarg;
		break;
	    case 'I':
		interfaceDir = optarg;
		break;
	    case 'o':
		executable = optarg;
		break;
	    case 'c':
		compiler = optarg;
		break;
	    case 'm':
		mainResName = optarg;
		break;
	    case 'v':
		verbose = TRUE;
		break;
	    case 'w':
		widelines = TRUE;
		break;
	    case 'C':
		compileOptions = optarg;
		break;
	    case 'L':
		linkerOptions = optarg;
		break;
	    case 'R':
		runtimeArgs = optarg;
		break;
	    case 'Z':
		zapFiles = optarg;
		break;
	    case '?':
	    default:
		usageMessage ();
		/* NOTREACHED */
	}

    if (verbose)
	trcexec = 1;

    if (!compiler) {		/* compiler name unspecified - use default */
	sprintf (tempString, "%s/sr", SRCMD);
	compiler = salloc (tempString);
    }

    setpath (interfaceDir ? interfaceDir : ".", 0); /* init for path searches */
    buildSymtab (argc, argv, optind);	/* construct the symbol table */
    if (!checkTable ())			/* check symbol table for errors */
	exit (1);
    if (verbose)
	printStats ();
    computeTransitiveClosure ();	/* compute complete dependency lists */
    writeMakefile (argv);		/* write out makefile */

    exit (0);
    /*NOTREACHED*/
}


usageMessage ()
{
    fprintf (stderr, 
"usage: srm [-{fIoc} path] [-m res] [-{CLRZ} args] [-{vw}] file.sr... [f.o...]\n");
    exit (1);
}


/* picks out a word from the given string and advances the string pointer to
 * the end of the word.  The word is extracted and returned.  Note the scanning
 * does not start at the start of the given string but one character after.
 */

char	*getarg (envStr)
char	*envStr;
{
    int		i, j;
    char	*retval;

    for (i = 1; * (envStr + i) == ' '; i++)
	;	/* skip over spaces */
    for (j = i; * (envStr + j) != ' ' && * (envStr + j) != '\0'; j++)
	;	/* got to end of word -- get word size */

    retval = alloc (j - i + 1);
    strncpy (retval, envStr + i, j - i);
    retval[j-i] = '\0';
    envStr += j;	/* set envStr to point to after the word */
    return retval;
}
